lib/segment/src/payload_storage/payload_storage.rs

```diff
@@ -1,17 +1,92 @@
 
 use crate::types::{PointOffsetType, PayloadKeyType, PayloadType, Filter, TheMap, PayloadSchemaType};
 use crate::entry::entry_point::OperationResult;
-
+use schemars::_serde_json::Value;
 
 /// Trait for payload data storage. Should allow filter checks
 pub trait PayloadStorage {
 
+    fn assign_all_with_value(&mut self, point_id: PointOffsetType, payload: TheMap<PayloadKeyType, serde_json::value::Value>) -> OperationResult<()> {
+        fn _extract_payloads<'a, I>(_payload: I, prefix_key: Option<PayloadKeyType>) -> Vec<(PayloadKeyType, Option<PayloadType>)>
+            where I: Iterator<Item=(&'a PayloadKeyType, &'a serde_json::value::Value)> {
+            _payload.flat_map(|(k, value)| {
+                let key = match &prefix_key {
+                    None => k.to_string(),
+                    Some(_k) => (_k.to_owned() + "__" + k).to_string(),
+                };
+                match value {
```

Comment:
serde supports parsing json Value into structures https://docs.serde.rs/serde_json/value/fn.from_value.html
This approach should also automatically handle complex structures like Geo payload. Current version will only flatten it into 2 float numbers.

And there are even a data structure which manages vector vs singular value interpretation:
https://github.com/qdrant/qdrant/blob/master/lib/collection/src/operations/payload_ops.rs#L30
You would probably want to move it into `segment` lib, I am ok with that.

The only thing which is not handled automatically is nested object. In this case I would suggest trying to convert the `Value` into `PayloadInterface` and if it fails - go one level deeper as it is currently implemented.

Reply:
For handling Geo Objects, I would suggest to have a convention that a nested object with some suffix like __geo__ to be treated as such.

This way you can handle all the patterns.

I had seen this PayloadInterface but I did not want to go outside segment lib. 

My question is if you would like to have this as the default payload method? or if u prefer to keep it as just an alternative

Reply:
> For handling Geo Objects, I would suggest to have a convention that a nested object with some suffix like geo to be treated as such.

Reply:
> For handling Geo Objects, I would suggest to have a convention that a nested object with some suffix like geo to be treated as such.

I would not do this, it breaks the existing API specification and can not be described in OpenAPI schema. The option with ready made `from_value` parsing already handles the case and is much easier to extend: you won't need to change the parsing function if some new data type is added. Pattern matching of PayloadInterface will ensure that no datatypes lost.

> I had seen this PayloadInterface but I did not want to go outside segment lib.

I can't see the problem with that, moving `PayloadInterface` into `segment` should not break anything. 

> My question is if you would like to have this as the default payload method? or if u prefer to keep it as just an alternative

The payload is parsed long before it is actually passed to segment library. So using of semi-parsed `Values` will introduce unnecessary complexity. So I prefer to keep this method as an alternative.

Reply:
I am not sure I understand your proposal, I cannot use `serde_json::from_value` for PayloadInterface Enums.

I am also not sure it would manage to cover the case where a `json` can be parsed automatically without the need to add the `type` tag like:

```json
"city": {"type": "keyword", "value": "Berlin"}},
```
while I think it would be cool to have:

```json
"city": {"Berlin"},
```

Reply:
> I am also not sure it would manage to cover the case where a json can be parsed automatically without the need to add the type tag like:

Reply:
> I am also not sure it would manage to cover the case where a json can be parsed automatically without the need to add the type tag like:

Similar results could be obtained with `#[serde(untagged)]` option, I will try to provide an example of the approach some later

Reply:
yes, thus I would need to change also the PayloadInterface definition? or would it be better to create an untagged version? or this can also be dynamically done when calling Serde?

Reply:
https://github.com/qdrant/qdrant/commit/ab822d72114e8f858e2a8147ce4415b7ba398a79 I would prefer this approach. It allows to use short and full versions simultaneously and also don't compromise the ability to extend with more complex data types.

Reply:
Thank you very much!

I updated the code with the new approach. As you suggest, I try to parse from value, and if not successfull I move to the next level.

I also added some tests for the parsing with and without strict from json.